src/funk/fd_funk_txn.c

```diff
@@ -0,0 +1,723 @@
+  txn->sibling_prev_cidx = fd_funk_txn_cidx( sibling_prev_idx     );
+  txn->sibling_next_cidx = fd_funk_txn_cidx( FD_FUNK_TXN_IDX_NULL );
+  txn->stack_cidx        = fd_funk_txn_cidx( FD_FUNK_TXN_IDX_NULL );
+  txn->tag               = 0UL;
+
+  /* Other dirty record tracking operations go here */
+
+  /* TODO: consider branchless impl */
+  if( FD_LIKELY( first_born ) ) *_child_head_cidx                         = fd_funk_txn_cidx( txn_idx ); /* opt for non-compete */
+  else                          map[ sibling_prev_idx ].sibling_next_cidx = fd_funk_txn_cidx( txn_idx );
+
+  *_child_tail_cidx = fd_funk_txn_cidx( txn_idx );
+
+  return txn;
+}
+
+/* fd_funk_txn_cancel_childless cancels a transaction that is known
+   to be childless.  Callers have already validated our input arguments.
+   Assumes that cancelling in the app can't fail but that could be
+   straightforward to support by giving this an error and plumbing
+   through to abort the overall cancel operation when it hits a error. */
+
+static void
+fd_funk_txn_cancel_childless( fd_funk_t *     funk,
+                              fd_funk_txn_t * map,
+                              ulong           txn_max,
+                              ulong           txn_idx ) {
+
+  /* Other cancel operations go here */
+
+  /* Leave the family */
+
+  ulong sibling_prev_idx = fd_funk_txn_idx( map[ txn_idx ].sibling_prev_cidx );
+  ulong sibling_next_idx = fd_funk_txn_idx( map[ txn_idx ].sibling_next_cidx );
+
+  /* TODO: Consider branchless impl */
+
+  if( FD_LIKELY( fd_funk_txn_idx_is_null( sibling_prev_idx ) ) ) { /* opt for non-compete */
+    ulong parent_idx = fd_funk_txn_idx( map[ txn_idx ].parent_cidx );
+    if( FD_LIKELY( fd_funk_txn_idx_is_null( parent_idx ) ) ) { /* No older sib and is a funk child, opt for incr pub */
+      funk->child_head_cidx = fd_funk_txn_cidx( sibling_next_idx );
+    } else { /* No older sib and has parent */
+      ASSERT_IN_PREP( parent_idx );
+      map[ parent_idx ].child_head_cidx = fd_funk_txn_cidx( sibling_next_idx );
+    }
+  } else { /* Has older sib */
+    ASSERT_IN_PREP( sibling_prev_idx );
+    map[ sibling_prev_idx ].sibling_next_cidx = fd_funk_txn_cidx( sibling_next_idx );
+  }
+
+  if( FD_LIKELY( fd_funk_txn_idx_is_null( sibling_next_idx ) ) ) { /* opt for non-compete */
+    ulong parent_idx = fd_funk_txn_idx( map[ txn_idx ].parent_cidx );
+    if( FD_LIKELY( fd_funk_txn_idx_is_null( parent_idx ) ) ) { /* No younger sib and is a funk child, opt for incr pub */
+      funk->child_tail_cidx = fd_funk_txn_cidx( sibling_prev_idx );
+    } else { /* No younger sib and has parent */
+      ASSERT_IN_PREP( parent_idx );
+      map[ parent_idx ].child_tail_cidx = fd_funk_txn_cidx( sibling_prev_idx );
+    }
+  } else { /* Has younger sib */
+    ASSERT_IN_PREP( sibling_next_idx );
+    map[ sibling_next_idx ].sibling_prev_cidx = fd_funk_txn_cidx( sibling_prev_idx );
+  }
+
+  fd_funk_txn_map_remove( map, &map[ txn_idx ].id );
```

Comment:
You have this form of remove in several places. Let's consider an alternate remove that takes "map + txn_idx" as the argument instead. Might be possible to optimize the map a bit for this case.

Reply:
Wrote some verbiage in fd_map_giant.h about that (might have blown it away before check in though).

In short, I agree.

In long, that's how the alternative maps already work and for similar reasons.  So when harmonizing the APIs between the new map_giant and the old maps, took a quick stab at doing this.  It was trickier than expected because the same that gives map_giant its properties that make it very useful here.

There were a bunch of possible implementations to do this.  The main thing is getting a way go from a map + idx to the original map in fast O(1) time with small O(1) memory.  The most straightforward is to have the elements themselves encode their index in the field directly.  Costs 8 bytes extra per element and didn't want to add the bloat given the design goals you described previously.  But the current layout with index compression has saved enough that we probably have room.

There are ways to do without any memory overhead at all in the element but they are a lot tougher to use practically (largely playing games with alignment and shared memory mappings) that are probably more mischievous than useful for a generic code generator.